rust/src/shards/gui/widgets/numeric_slider.rs

```diff
@@ -0,0 +1,276 @@
+              if var.exposedType.basicType != t.basicType {
+                return Err(concat!($name_str, ": incorrect type of variable."));
+              }
+              break;
+            }
+          }
+        }
+
+        Ok(common_type::bool)
+      }
+
+      fn exposedVariables(&mut self) -> Option<&ExposedTypes> {
+        if self.variable.is_variable() && self.should_expose {
+          self.exposing.clear();
+
+          let exp_info = ExposedInfo {
+            exposedType: common_type::$type,
+            name: self.variable.get_name(),
+            help: cstr!("The exposed variable").into(),
+            ..ExposedInfo::default()
+          };
+
+          self.exposing.push(exp_info);
+          Some(&self.exposing)
+        } else {
+          None
+        }
+      }
+
+      fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+        self.requiring.clear();
+
+        // Add UI.Parents to the list of required variables
+        let exp_info = ExposedInfo {
+          exposedType: EGUI_UI_SEQ_TYPE,
+          name: self.parents.get_name(),
+          help: cstr!("The parent UI objects.").into(),
+          ..ExposedInfo::default()
+        };
+        self.requiring.push(exp_info);
+
+        Some(&self.requiring)
+      }
+
+      fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+        self.parents.warmup(ctx);
+        self.label.warmup(ctx);
+        self.variable.warmup(ctx);
+        self.min.warmup(ctx);
+        self.max.warmup(ctx);
+
+        if self.should_expose {
+          self.variable.get_mut().valueType = common_type::$type.basicType;
+        }
+
+        Ok(())
+      }
+
+      fn cleanup(&mut self) -> Result<(), &str> {
+        self.max.cleanup();
+        self.min.cleanup();
+        self.variable.cleanup();
+        self.label.cleanup();
+        self.parents.cleanup();
+
+        Ok(())
+      }
+
+      fn activate(&mut self, _context: &Context, _input: &Var) -> Result<Var, &str> {
+        if let Some(ui) = util::get_current_parent(*self.parents.get())? {
+          let value = &mut if self.variable.is_variable() {
+            unsafe { self.variable.get_mut().payload.__bindgen_anon_1.$field }
+          } else {
+            self.tmp
+          };
+          let max = self.max.get().try_into()?;
+          let min = self.min.get().try_into()?;
+
+          let mut slider = egui::Slider::new(value, min..=max);
+          if !self.label.get().is_none() {
+            let text: &str = self.label.get().try_into()?;
+            slider = slider.text(text);
+          }
+
+          let response = ui.add(slider);
+          if response.changed() || response.lost_focus() {
+            if self.variable.is_variable() {
+              self.variable.set((*value).into());
+            } else {
+              self.tmp = *value;
+            }
+          }
+          Ok((*value).into())
+        } else {
+          Err("No UI parent")
+        }
+      }
+    }
+  };
+}
+
+impl_ui_slider!(
```

Comment:
What's the reason for having these be separate shards, instead of just UI.Slider and passing either and int or float variable

Reply:
@Kryptos-FR , @guusw is right. This is a bit lazy 😭
The type system is indeed there to avoid this scenario.

Reply:
Good point. I'll have a look.

Reply:
You know it in compose.

Reply:
We have to keep two separate shards. That's because we can't otherwise know the type of the expsoed variable.